### ISSUE TITLE
aws-lambda: Function.grantInvoke() not applying permissions for older versions of lambda function

### ISSUE THREAD

**Issue by michaellasmanis** [aws/aws-cdk#32455]
### Describe the bug

when using Function.grantInvoke(), permission for function:Invoke is applied to the unqualified function arn but not to any of the older versions. 

### Regression Issue

- [ ] Select this option if this issue appears to be a regression.

### Expected Behavior

Per the documentation here: https://docs.aws.amazon.com/cdk/api/v2/java/software/amazon/awscdk/services/lambda/package-summary.html#resource-based-policies-heading

and the GitHub PR here: https://github.com/aws/aws-cdk/pull/19318

I would expect that both the unqualified [FunctionArn] and [Function.Arn].* to have the permissions applied.

### Current Behavior

Invoke permission only seem to be applied to the unqualified function arn.

cdk synth produces the following snippet:

```
  HelloWorldFunctionInvokeFcyXBRX02EWa52GlFECQiCzDt0fdRUDi4mo4foC5aU2283F0E1:
    Type: AWS::Lambda::Permission
    Properties:
      Action: lambda:InvokeFunction
      FunctionName:
        Fn::GetAtt:
          - HelloWorldFunctionB2AB6E79
          - Arn
      Principal: apigateway.amazonaws.com

```

which just enabled the unqualified ARN

### Reproduction Steps

if you have a stack containing such:

```
    // Define the Lambda function resource
    const myFunction = new lambda.Function(this, "HelloWorldFunction", {
      runtime: lambda.Runtime.NODEJS_20_X, // Provide any supported Node.js runtime
      handler: "index.handler",
      code: lambda.Code.fromInline(`
        exports.handler = async function(event) {
          return {
            statusCode: 200,
            body: JSON.stringify('Hello World on ` + new Date().toLocaleString() + `!'),
          };
        };
      `),
    });

    const myVersion = new lambda.Version(this, "HelloWorldFunction" + new Date().getTime()/1000,
    {
      lambda: myFunction,
      removalPolicy: cdk.RemovalPolicy.RETAIN_ON_UPDATE_OR_DELETE
    });

    myFunction.grantInvoke(new iam.ServicePrincipal('apigateway.amazonaws.com'));
```

you will notice after the second (or subsequent) deploy that the permissions from the non-$LATEST versions are removed as part of the cdk deploy.
 

### CDK CLI Version

2.172.0

### Node.js Version

v23.3.0

### OS

macOS

### Language

Java

### Language Version

java 17

**Comment by michaellasmanis** [comment]
sample project based on cdk init
[hello-cdk-java.zip](https://github.com/user-attachments/files/18071244/hello-cdk-java.zip)

**Comment by nmussy** [comment]
Hey @michaellasmanis,

There definitely is a difference between granting invoke to a role vs a service. The former is applied directly to the function's policy document, and the latter is applied via a [`CfnPermission`](https://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/aws-resource-lambda-permission.html) resource . This is the current behavior:

```ts
// GIVEN
const stack = new cdk.Stack();
const role = new iam.Role(stack, 'Role', {
  assumedBy: new iam.AccountPrincipal('1234'),
});
const service = new iam.ServicePrincipal('apigateway.amazonaws.com');
const fn = new lambda.Function(stack, 'Function', {
  code: lambda.Code.fromInline('xxx'),
  handler: 'index.handler',
  runtime: lambda.Runtime.NODEJS_LATEST,
});

// WHEN
fn.grantInvoke(role);
fn.grantInvoke(service);

// THEN
expect(fn.resourceArnsForGrantInvoke).toEqual([fn.functionArn, `${fn.functionArn}:*`]);
Template.fromStack(stack).hasResourceProperties('AWS::IAM::Policy', {
  PolicyDocument: {
    Version: '2012-10-17',
    Statement: [
      {
        Action: 'lambda:InvokeFunction',
        Effect: 'Allow',
        Resource: [
          { 'Fn::GetAtt': ['Function76856677', 'Arn'] },
          { 'Fn::Join': ['', [{ 'Fn::GetAtt': ['Function76856677', 'Arn'] }, ':*']] },
        ],
      },
    ],
  },
});
// There is only a single Permission resource
Template.fromStack(stack).resourceCountIs('AWS::Lambda::Permission', 1);
Template.fromStack(stack).hasResourceProperties('AWS::Lambda::Permission', {
  Action: 'lambda:InvokeFunction',
  FunctionName: { 'Fn::GetAtt': ['Function76856677', 'Arn'] },
  Principal: 'apigateway.amazonaws.com',
});
```

It seems like you should be able to qualify [`CfnPermission.FunctionName`](https://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/aws-resource-lambda-permission.html#cfn-lambda-permission-functionname), but I haven't looked too deeply into it yet

**Comment by michaellasmanis** [comment]
@nmussy 

Thanks for the clarification, that makes sense.  However, there is then definitely a documentation error at least in the javadocs:
[https://docs.aws.amazon.com/cdk/api/v2/java/software/amazon/awscdk/services/lambda/package-summary.html#resource-based-policies-heading](https://docs.aws.amazon.com/cdk/api/v2/java/software/amazon/awscdk/services/lambda/package-summary.html#resource-based-policies-heading
)
```
 ServicePrincipal principal = new ServicePrincipal("my-service");
 
 fn.grantInvoke(principal);

```
> 
> By default fn.grantInvoke() grants permission to the principal to invoke any version of the function, including all past ones. If you only want the principal to be granted permission to invoke the latest version or the unqualified Lambda ARN, use grantInvokeLatestVersion(grantee).
> 

The example code from the javadocs is basically what I copied and pasted into my test case.

Basically, I'm trying to work around issue #28412.  I'll check out [CfnPermission.FunctionName](https://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/aws-resource-lambda-permission.html#cfn-lambda-permission-functionname) to see if it can work for me.  Any other suggestions are highly appreciated.

**Comment by nmussy** [comment]
I think there's an issue with the current implementation of the internal `grant` method when given a service service, as it seems the qualifier is being ignored even when explicitly set:

```ts
const fn = new lambda.Function(stack, 'Function', {
  code: lambda.Code.fromInline('xxx'),
  handler: 'index.handler',
  runtime: lambda.Runtime.NODEJS_LATEST,
});
const service = new iam.ServicePrincipal('apigateway.amazonaws.com');

fn.grantInvoke(service);
fn.grantInvokeLatestVersion(service);
fn.grantInvokeVersion(service, fn.latestVersion);
```

All 3 grants generate effectively the same policy:

```json
[
    "FunctionInvokeFcyXBRX02EWa52GlFECQiCzDt0fdRUDi4mo4foC5aU730270F4": {
      "Type": "AWS::Lambda::Permission",
      "Properties": {
        "Action": "lambda:InvokeFunction",
        "FunctionName": {
          "Fn::GetAtt": [
            "Function76856677",
            "Arn"
          ]
        },
        "Principal": "apigateway.amazonaws.com"
      }
    },
    "FunctionInvokeKTHGXOMcvOrpWFkuCdKIjgLLIicmjeE3TRCaRos63A32ADB": {
      "Type": "AWS::Lambda::Permission",
      "Properties": {
        "Action": "lambda:InvokeFunction",
        "FunctionName": {
          "Fn::GetAtt": [
            "Function76856677",
            "Arn"
          ]
        },
        "Principal": "apigateway.amazonaws.com"
      }
    },
    "FunctionCurrentVersion4E2B2261ecf75ed18d6b6a0a5635c08929baf432": {
      "Type": "AWS::Lambda::Version",
      "Properties": {
        "FunctionName": {
          "Ref": "Function76856677"
        }
      }
    },
    "FunctionInvokewPiaqpkwMJl8Ed9zUwcfGrfRum0xC0KM1xqhfffIXlo2E3F6599": {
      "Type": "AWS::Lambda::Permission",
      "Properties": {
        "Action": "lambda:InvokeFunction",
        "FunctionName": {
          "Fn::GetAtt": [
            "Function76856677",
            "Arn"
          ]
        },
        "Principal": "apigateway.amazonaws.com"
      }
    }
  }
]
```

**Comment by michaellasmanis** [comment]
I agree, I would expect fn.grantInvokeVersion(service, fn.latestVersion) to create a qualified permission.  I just tested with an explicitly created version and it had the same effect:

```
    const fn = new lambda.Function(this, 'Function', {
      code: lambda.Code.fromInline('xxx'),
      handler: 'index.handler',
      runtime: lambda.Runtime.NODEJS_LATEST,
    });
    const service = new iam.ServicePrincipal('apigateway.amazonaws.com');

    const myVersion = new lambda.Version(this, "HelloWorldFunction" + new Date().getTime()/1000,
    {
      lambda: fn
    });

    fn.grantInvoke(service);
    fn.grantInvokeLatestVersion(service);
    fn.grantInvokeVersion(service, myVersion);
  }

```

**Comment by khushail** [comment]
@michaellasmanis , thanks for reporting this issue and your analysis reg the code execution.
@nmussy , appreciate your deep dive analysis and sharing the insights.

I agree with @michaellasmanis that the corresponding documentation should clarify or be updated with the actual behavior- https://docs.aws.amazon.com/cdk/api/v2/docs/aws-cdk-lib.aws_lambda-readme.html#grant-function-access-to-aws-services in all supported languages(Typescript/java/python etc.) 
It also makes a fair case to address this bug as well as pointed out by @nmussy  [here](https://github.com/aws/aws-cdk/issues/32455#issuecomment-2533788776).

I would also be reaching out to Team if its something that should be considered and share their insights here if possible.

Thanks.
